acl/aclauthz/acl_cache.go

```diff
@@ -264,3 +266,54 @@ func (c *ACLCaches) Purge() {
 		}
 	}
 }
+
+// TODO: unexport once this is moved into the same package as ACLResolver
+func NewPolicyAuthorizerWithCache(policies structs.ACLPolicies, cache *ACLCaches, entConf *acl.Config) (acl.Authorizer, error) {
+	// Determine the cache key
+	cacheKey := policies.HashKey()
+	entry := cache.GetAuthorizer(cacheKey)
+	if entry != nil {
+		// the hash key takes into account the policy contents. There is no reason to expire this cache or check its age.
+		return entry.Authorizer, nil
+	}
+
+	parsed, err := resolveWithCache(policies, cache, entConf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse the ACL policies: %v", err)
+	}
+
+	// Create the ACL object
+	authorizer, err := acl.NewPolicyAuthorizer(parsed, entConf)
+	if err != nil {
+		return nil, fmt.Errorf("failed to construct ACL Authorizer: %v", err)
+	}
+
+	// Update the cache
+	cache.PutAuthorizer(cacheKey, authorizer)
+	return authorizer, nil
+}
+
+func resolveWithCache(policies structs.ACLPolicies, cache *ACLCaches, entConf *acl.Config) ([]*acl.Policy, error) {
```

Comment:
This was previously `ACLPolicies.resolveWithCache`, the only change is the function signature